### ISSUE TITLE
Huge bundle size

### ISSUE THREAD

**Issue by Hurtak** [airbnb/react-outside-click-handler#24]
This library is 36KB (after compile + minify) which is a lot and seems order of magnitude bigger than what it should be. One of its subdependencies is probably at fault here, have you considered removing them? For example Object.values has decent support now and it should be library users responsibility to polyfill std lib functions.

**Comment by ljharb** [comment]
The default is the safe thing, to include it - you can easily alias it out of your bundle if you know you’re polyfilling it or if you only support browsers that do (even modern browsers often require polyfills, and Object.values is new).

In other words, it is not the library user’s responsibility to polyfill; it is always best when the library makes that unnecessary.

**Comment by ljharb** [comment]
What is the user’s responsibility and not the library's, however, is the resulting bundle size :-)

**Comment by Hurtak** [comment]
Object.values was just an example, I did not actually look into the subdependencies sizes. But on the other hand in your readme you mention "For full browser support, react-outside-click-handler expects its consumer to include airbnb-browser-shims or another similar shim. This is specifically due to the use of the node.contains method." so if user is expected to polyfill node.contains why he is not expected to polyfill Object.values?

One other outside-click library is around 5KB and it does seem to have about the same lines of source code so it seemed strange on first look that this library is 7 times bigger while it does not provide multiples of funcionality (I might be wrong).

"What is the user’s responsibility and not the library's, however, is the resulting bundle size :-)" Which is affected a lot by its dependencies, that is kidna the point of this issue :)

TLDR: point of this ticket was that library size seems big on first look, so it might be worth a while to investigate its dependencies and maybe replacing them (eg object.values with just regular Object.values or consolidated-events with native addEventListener).

**Comment by ljharb** [comment]
Hmm, that is a good point. consolidated-events, for one, is still needed because addEventListener can’t be fully polyfilled, but there might be some value in reevaluating dependencies and polyfill requirements. I’ll take a look.

(What i meant was, as the end user you can apply all sorts of optimizations in your bundler that can’t be safely applied in the general case, like “replace all object-values imports with Object.values”)

**Comment by ljharb** [comment]
I've gone ahead and removed that requirement from the readme in 4e993590c16ffc4184101317c81aa9fdbe37e6b9 by using a separate `document.contains` package instead of requiring the polyfill; as such, now you can alias out the various deps like `object.values`, `document.contains`, etc, in your bundler, when you have the ability to know that your supported browsers don't need them.

**Comment by Hurtak** [comment]
"as the end user you can apply all sorts of optimizations in your bundler" This heavily depends on framework/bundler user is using. For example I am not sure if some other bundler than webpack offer this option (aliasing module). Also there are frameworks like Create React App that do not expose these parts of webpack config.

**Comment by ljharb** [comment]
Every bundler I’m aware of does; webpack got it from browserify. As for CRA, they’ve chosen to require ejecting to achieve anything really useful, at which point you can customize it all you like. Another alternative would be for them to configure this aliasing for everyone.

Either way, if you care about your bundle sizes, you need to be using a bundler that provides maximal configurability - CRA isn’t that.

**Comment by oliviertassinari** [comment]
People can find the detail of the bundle size of the package in https://bundlephobia.com/result?p=react-outside-click-handler.

**Comment by ljharb** [comment]
@oliviertassinari that’s an extremely misleading site, because es-abstract is probably already in most bundles, since it’s a dep of basically all the non-core-js shims; and if you follow the advice above, you can alias it out anyways.

I think this has been sufficiently answered.

**Comment by oliviertassinari** [comment]
@ljharb I didn't know that :)
